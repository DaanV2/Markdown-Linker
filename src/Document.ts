import * as fs from "fs";

export async function LoadDocument(doc: string): Promise<string> {
  return fs.promises.readFile(doc, "utf8");
}

export async function SaveDocument(doc: string, content: string): Promise<void> {
  return fs.promises.writeFile(doc, content, "utf8");
}
