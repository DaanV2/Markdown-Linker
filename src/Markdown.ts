export function isInLink(text: string, offset: number): boolean {
  const end = text.length;

  // Look for the ](
  for (let i = offset; i <= end; i++) {
    if (text[i] === "]") {
      if (text[i + 1] === "(") {
        return true;
      }

      return false;
    }
  }

  return false;
}
