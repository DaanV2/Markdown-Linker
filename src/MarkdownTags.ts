export interface MarkdownTags {
  tags: Array<string>;
}

export namespace MarkdownTags {
  export function is(value: any): value is MarkdownTags {
    return typeof value === "object" && value !== null && Array.isArray(value.tags);
  }
}
