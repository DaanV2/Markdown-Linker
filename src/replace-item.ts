export interface ReplaceItem {
  tag: string;
  url: string;
  search: RegExp;
  tagSearch: RegExp;
}
