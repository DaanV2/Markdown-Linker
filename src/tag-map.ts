import { loadDocument } from "./document";
import { MarkdownTags, SerializedTagMap } from "./markdown-tags";
import { ReplaceItem } from "./replace-item";
import * as core from "@actions/core";
import FastGlob from "fast-glob";

export class TagMap {
  private _data: ReplaceItem[];

  constructor() {
    this._data = [];
  }

  public get count() {
    return this._data.length;
  }

  public add(tags: string | Array<string>, uri: string, search?: string): this {
    if (!Array.isArray(tags)) tags = [tags];

    tags.forEach((tag) => {
      this._data.push({
        tag: tag,
        url: uri,
        search: new RegExp(search || `(?:^|(?:[^#]|^)[\\s,.!?])(${tag})(?:$|[\\s,.!?])`, "gmi"),
        tagSearch: new RegExp(tag, "gmi"),
      });
    });

    return this;
  }

  public rewriteLinks(content: string): string {
    const linkDefinitions: Record<string, string> = {};
    const linkRegex = /\[([^\]]+)\]\(([^)]+)\)/g;
    
    // Collect existing link definitions
    const existingLinkDefs = content.match(/^\[([^\]]+)\]:\s*(.+)$/gm) || [];
    existingLinkDefs.forEach((def) => {
      const parts = def.match(/^\[([^\]]+)\]:\s*(.+)$/);
      if (parts) {
        linkDefinitions[parts[1].toLowerCase()] = parts[2];
      }
    });
  
    // Rewrite inline links to reference links
    content = content.replace(linkRegex, (_fullMatch, text, url) => {
      let ref = Object.keys(linkDefinitions).find((key) => linkDefinitions[key] === url);
      if (!ref) {
        ref = `${text}`;
        linkDefinitions[ref] = url;
      }
      return `[${text}]`;
    });
  
    // Append updated link definitions to the bottom
    const linkDefs = Object.entries(linkDefinitions)
      .map(([key, url]) => `[${key}]: ${url}`)
      .join("\n");
  
    return `${content}\n\n${linkDefs}`;
  }

  public has(tag: string): boolean {
    return this._data.some((item) => item.tag === tag);
  }

  public forEach(callbackfn: (item: ReplaceItem) => void, thisArg?: any) {
    return this._data.forEach(callbackfn, thisArg ?? this);
  }

  public async scrapeDoc(doc: string): Promise<void> {
    try {
      const content = await loadDocument(doc);
      return this.scrape(content, doc);
    } catch (err) {
      console.warn(`Error scraping ${doc}`, err);
      core.notice(`Error scraping ${doc}\n${err}`, { file: doc });
    }
  }

  public scrape(content: string, doc: string) {
    const match = content.match(/<!--{.*}-->/gim);

    if (match) {
      for (const item of match) {
        const data = item.replace(/<!--|-->/gim, "");

        if (data.startsWith("{") && data.endsWith("}")) {
          const json = JSON.parse(data);

          if (MarkdownTags.is(json)) {
            this.add(json.tags, doc);
          }
        }
      }
    }
  }

  public async loadMap(filepath: string): Promise<void> {
    try {
      console.log("Loading map", filepath);
      const content = await loadDocument(filepath);
      const json = JSON.parse(content);

      //Convert array of tags to a map
      if (Array.isArray(json)) {
        json.forEach((item) => {
          if (SerializedTagMap.is(item)) {
            this.add(item.tag, item.url, item.search);
          }
        });
      }
    } catch (err) {
      console.warn("Error loading tag map", err);
      core.notice(`Error loading tag map ${filepath}\n${err}`, { file: filepath });
    }
  }

  public async loadMaps(dir: string): Promise<void> {
    const files = FastGlob.sync(["**/.tags/*.json"], { cwd: dir, absolute: true });

    return Promise.all(files.map((file) => this.loadMap(file))).then(() => {});
  }
}
