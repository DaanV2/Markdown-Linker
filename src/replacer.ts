import { isInLink } from "./markdown";
import { loadDocument, saveDocument } from "./document";
import { TagMap } from "./tag-map";
import path from "path";

export class Replacer {
  private tags: TagMap;

  constructor(tags: TagMap) {
    this.tags = tags;
  }

  public async replaceDocuments(files: string[]): Promise<void> {
    await Promise.all(files.map((file) => this.replaceDocument(file)));
  }

  public async replaceDocument(file: string): Promise<void> {
    const content = await loadDocument(file);
    const newContent = this.replace(content, file);

    if (content !== newContent) {
      console.info("Saving", file);
      return saveDocument(file, newContent);
    }
  }

  public replace(text: string, currentDoc: string): string {
    const lines = text.split(/\n/);
    const max = lines.length;

    this.tags.forEach((item) => {
      let { tag, url, tagSearch, search } = item;
      if (url === currentDoc) return;
      const dir = path.resolve(currentDoc, "..");

      if (!url.startsWith("http")) {
        url = path.relative(dir, url).replace(/\\/gim, "/");
        url = encodeURI(url);
      }

      for (let i = 0; i < max; i++) {
        let line = lines[i];

        // Skip headers
        if (line.startsWith("#")) continue;
        //If its pure html, skip it
        if (line.match(/<[^>]*>/gim)) continue;
        // Skip code blocks
        if (line.startsWith("```")) {
          //Find end of code block
          line = lines[++i];
          while (i < max && !line.startsWith("```")) {
            i++;
            line = lines[i];
          }
          continue;
        }

        lines[i] = line.replace(search, (match: string, tag, offset, original) => {
          if (isInLink(original, offset)) return match;
          const newText = `[${tag}](${url})`;

          return match.replace(tagSearch, newText);
        });
      }
    });

    return lines.join("\n");
  }
}
