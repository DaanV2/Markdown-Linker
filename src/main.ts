import { Replacer } from "./replacer";
import { TagMap } from "./tag-map";
import FastGlob from "fast-glob";
import * as core from "@actions/core";

export interface Settings {
  folder: string;
  ignores: string[];
}

export async function main(settings: Settings) {
  const cwd = settings.folder;
  const tagMap = new TagMap();

  // Loading maps
  await core.group("Loading maps", async () => {
    return tagMap.loadMaps(cwd);
  });

  // Gathering files
  const files = await core.group("Gathering files", async () => {
    const ignores = ["node_modules", ".tags", ...settings.ignores];
    return FastGlob.sync(["**/*.md"], { cwd, absolute: true, ignore: ignores });
  });

  // Scraping files
  await core.group("Scraping files", async () => {
    return Promise.all(files.map((file) => tagMap.scrapeDoc(file)));
  });
  console.info("Tag count:", tagMap.count);

  // Replacing tags
  await core.group("Replacing tags", async () => {
    const replacer = new Replacer(tagMap);
    return replacer.replaceDocuments(files);
  });
}
