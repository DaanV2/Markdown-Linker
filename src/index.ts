import { main } from "./main";
import * as core from "@actions/core";

async function action() {
  const settings = {
    folder: core.getInput("folder", { trimWhitespace: true }),
    ignores: core.getMultilineInput("ignore", { trimWhitespace: true }),
  };

  console.log("Starting on linking", settings);
  
  await main(settings);
}

action()
  .then(() => console.log("done"))
  .catch((err) => core.setFailed(err));
