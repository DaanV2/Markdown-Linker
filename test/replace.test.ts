import { Replacer } from "../src/replacer";
import { TagMap } from "../src/tag-map";
import path from "path";

describe("Scraping", () => {
  test("Able to scrape the content", () => {
    const lines = `
    # Example Markdown Heading
    <!--{"tags":["Frodo baggings"]}-->
    Some text`;

    const tagMap = new TagMap();
    tagMap.scrape(lines, "C:/test.md");

    expect(tagMap.has("Frodo baggings")).toBe(true);
  });
});

describe("Replacing", () => {
  const root = __dirname;
  const folder = path.join(root, "..", "example");
  const filepath = path.join(folder, "test this.md");
  const otherFilepath = path.join(folder, "other.md");

  const tags = new TagMap();
  tags.add("Frodo baggings", otherFilepath);
  const replacer = new Replacer(tags);

  const tests: { from: string; to: string }[] = [
    {
      from: "Frodo baggings",
      to: `[Frodo baggings](other.md)`,
    },
    {
      from: " Frodo baggings ",
      to: ` [Frodo baggings](other.md) `,
    },
    {
      from: "Frodo baggings is a great guy",
      to: `[Frodo baggings](other.md) is a great guy`,
    },
    {
      from: "Frodo baggings is a great guy",
      to: `[Frodo baggings](other.md) is a great guy`,
    },
    {
      from: "Is Frodo baggings a great guy?",
      to: `Is [Frodo baggings](other.md) a great guy?`,
    },
    {
      from: "# Frodo baggings",
      to: "# Frodo baggings",
    },
    {
      from: "##### Frodo baggings",
      to: "##### Frodo baggings",
    },
    {
      from: "Is [Frodo baggings](../example/mark.md) a great guy?",
      to: "Is [Frodo baggings](../example/mark.md) a great guy?",
    },
    {
      from: "Is [Guy Frodo baggings](../example/mark.md) a great guy?",
      to: "Is [Guy Frodo baggings](../example/mark.md) a great guy?",
    },
  ];

  tests.forEach((t) => {
    test(`Able to replace '${t.from}' with '${t.to}'`, () => {
      const result = replacer.replace(t.from, filepath);

      expect(result).toMatchSnapshot();
    });
  });
});
