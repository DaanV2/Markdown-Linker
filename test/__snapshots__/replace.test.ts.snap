// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Replacing Able to replace ' Frodo baggings ' with ' [Frodo baggings](other.md) ' 1`] = `
" [Frodo baggings] 

[Frodo baggings]: other.md"
`;

exports[`Replacing Able to replace '# Frodo baggings' with '# Frodo baggings' 1`] = `"# Frodo baggings"`;

exports[`Replacing Able to replace '##### Frodo baggings' with '##### Frodo baggings' 1`] = `"##### Frodo baggings"`;

exports[`Replacing Able to replace 'Frodo baggings is a great guy' with '[Frodo baggings](other.md) is a great guy' 1`] = `
"[Frodo baggings] is a great guy

[Frodo baggings]: other.md"
`;

exports[`Replacing Able to replace 'Frodo baggings is a great guy' with '[Frodo baggings](other.md) is a great guy' 2`] = `
"[Frodo baggings] is a great guy

[Frodo baggings]: other.md"
`;

exports[`Replacing Able to replace 'Frodo baggings' with '[Frodo baggings](other.md)' 1`] = `
"[Frodo baggings]

[Frodo baggings]: other.md"
`;

exports[`Replacing Able to replace 'Is [Frodo baggings](../example/mark.md) a great guy?' with 'Is [Frodo baggings](../example/mark.md) a great guy?' 1`] = `
"Is [Frodo baggings] a great guy?

[Frodo baggings]: ../example/mark.md"
`;

exports[`Replacing Able to replace 'Is [Guy Frodo baggings](../example/mark.md) a great guy?' with 'Is [Guy Frodo baggings](../example/mark.md) a great guy?' 1`] = `
"Is [Guy Frodo baggings] a great guy?

[Guy Frodo baggings]: ../example/mark.md"
`;

exports[`Replacing Able to replace 'Is Frodo baggings a great guy?' with 'Is [Frodo baggings](other.md) a great guy?' 1`] = `
"Is [Frodo baggings] a great guy?

[Frodo baggings]: other.md"
`;
